### ISSUE TITLE
Relay + PHP 7.4 / 8.1 + paratest causes Exit Code: 139(Segmentation violation) 

### ISSUE THREAD

**Issue by dorrogeray** [cachewerk/relay#83]
![image](https://github.com/cachewerk/relay/assets/4999006/031da321-1088-4cbf-8de3-92265ced7975)

## Description
Our integration tests work fine without relay extension installed. With it, the failure is random.

## Expected Behavior
No Exit Code: 139(Segmentation violation) should occur

## Actual Behavior
Exit Code: 139(Segmentation violation) happens randomly

## Possible Fix
Stability

## Steps to Reproduce
Can't really provide exact repro steps - our integration suite is large and so just the amount of tests may play a role.

## Additional context
We are not even using the relay actively in the codebase yet. It is just installed.

## Relay diagnostics

```

relay

Relay Support => enabled
Relay Version => 0.6.5
Available cache => 160
Available serializers => php, json, igbinary, msgpack
Available compression => lzf, zstd, lz4
Binary UUID => c9a56f96-8233-49a8-bc03-b5e91d72b8c9
Git SHA => 6c2bffbe43d7da174f50647ac6790be55da64cb4
Allocator => relay
License state => unknown
License memory cap => 0
License request id =>  
relay.enabled => false
relay.key =>  
relay.maxmemory => 0
relay.maxmemory_pct => 95
relay.eviction_policy => noeviction
relay.eviction_sample_keys => 128
relay.initial_readers => 32
relay.invalidation_poll_usec => 5
relay.pconnect_default => 1
relay.max_endpoint_dbs => 32
relay.loglevel => debug
relay.logfile => /var/log/relay.log
```
logs
```
[1687437220.168872 DBG 8] shmalloc.c:202 Allocated 8192 byte pool (head: 7f21e2201000, curr: 7f21e2202020, alignment: 16)
[1687437220.168904 NOT 8] relay.c:2256 STARTUP - Relay has started...
[1687437595.821045 DBG 120] shmalloc.c:202 Allocated 8192 byte pool (head: 7f129b401000, curr: 7f129b402020, alignment: 16)
[1687437595.821070 NOT 120] relay.c:2256 STARTUP - Relay has started...
[1687437595.821504 DBG 120] relay.c:2911 RINIT - Requests (1 total, 1 active, 1 max)
[1687437595.821617 DBG 120] relay.c:2878 HTTP_HOST - Failed to find HTTP_HOST or it's not a string.
[1687437595.821620 DBG 120] relay.c:2854 PLINK - Uninstalling handlers for 0 links
[1687437595.821967 DBG 120] cache.c:303 Failed to get global epoch
[1687437595.821973 DBG 120] relay.c:3003 RSHUTDOWN - Finished request
[1687437595.822264 NOT 120] relay.c:2279 Relay is shutting down, bye bye
[1687437625.020953 DBG 122] shmalloc.c:202 Allocated 8192 byte pool (head: 7fba85a00000, curr: 7fba85a01020, alignment: 16)
[1687437625.020979 NOT 122] relay.c:2256 STARTUP - Relay has started...
[1687437625.021435 DBG 122] relay.c:2911 RINIT - Requests (1 total, 1 active, 1 max)
[1687437625.021540 DBG 122] relay.c:2878 HTTP_HOST - Failed to find HTTP_HOST or it's not a string.
[1687437625.021543 DBG 122] relay.c:2854 PLINK - Uninstalling handlers for 0 links
[1687437627.220998 DBG 128] shmalloc.c:202 Allocated 8192 byte pool (head: 7f2bb4202000, curr: 7f2bb4203020, alignment: 16)
[1687437627.221024 NOT 128] relay.c:2256 STARTUP - Relay has started...
[1687437627.221544 DBG 128] relay.c:2911 RINIT - Requests (1 total, 1 active, 1 max)
[1687437627.221657 DBG 128] relay.c:2878 HTTP_HOST - Failed to find HTTP_HOST or it's not a string.
[1687437627.221660 DBG 128] relay.c:2854 PLINK - Uninstalling handlers for 0 links
[1687437627.223597 DBG 129] shmalloc.c:202 Allocated 8192 byte pool (head: 7fbb4f000000, curr: 7fbb4f001020, alignment: 16)
[1687437627.223622 NOT 129] relay.c:2256 STARTUP - Relay has started...
[1687437627.224151 DBG 129] relay.c:2911 RINIT - Requests (1 total, 1 active, 1 max)
[1687437627.224234 DBG 129] relay.c:2878 HTTP_HOST - Failed to find HTTP_HOST or it's not a string.
[1687437627.224237 DBG 129] relay.c:2854 PLINK - Uninstalling handlers for 0 links
[1687437627.228531 DBG 130] shmalloc.c:202 Allocated 8192 byte pool (head: 7f9f7e401000, curr: 7f9f7e402020, alignment: 16)
[1687437627.228554 NOT 130] relay.c:2256 STARTUP - Relay has started...
[1687437627.229048 DBG 130] relay.c:2911 RINIT - Requests (1 total, 1 active, 1 max)
[1687437627.229149 DBG 130] relay.c:2878 HTTP_HOST - Failed to find HTTP_HOST or it's not a string.
[1687437627.229152 DBG 130] relay.c:2854 PLINK - Uninstalling handlers for 0 links
[1687437627.230938 DBG 131] shmalloc.c:202 Allocated 8192 byte pool (head: 7f5e49000000, curr: 7f5e49001020, alignment: 16)
[1687437627.230960 NOT 131] relay.c:2256 STARTUP - Relay has started...
[1687437627.231399 DBG 131] relay.c:2911 RINIT - Requests (1 total, 1 active, 1 max)
[1687437627.231495 DBG 131] relay.c:2878 HTTP_HOST - Failed to find HTTP_HOST or it's not a string.
[1687437627.231498 DBG 131] relay.c:2854 PLINK - Uninstalling handlers for 0 links
[1687437627.236027 DBG 132] shmalloc.c:202 Allocated 8192 byte pool (head: 7fc1c601e000, curr: 7fc1c601f020, alignment: 16)
[1687437627.236027 DBG 133] shmalloc.c:202 Allocated 8192 byte pool (head: 7f15275e5000, curr: 7f15275e6020, alignment: 16)
[1687437627.236046 NOT 132] relay.c:2256 STARTUP - Relay has started...
[1687437627.236048 NOT 133] relay.c:2256 STARTUP - Relay has started...
[1687437627.236502 DBG 132] relay.c:2911 RINIT - Requests (1 total, 1 active, 1 max)
[1687437627.236502 DBG 133] relay.c:2911 RINIT - Requests (1 total, 1 active, 1 max)
[1687437627.236574 DBG 133] relay.c:2878 HTTP_HOST - Failed to find HTTP_HOST or it's not a string.
[1687437627.236577 DBG 133] relay.c:2854 PLINK - Uninstalling handlers for 0 links
[1687437627.236598 DBG 132] relay.c:2878 HTTP_HOST - Failed to find HTTP_HOST or it's not a string.
[1687437627.236601 DBG 132] relay.c:2854 PLINK - Uninstalling handlers for 0 links
[1687437627.239944 DBG 134] shmalloc.c:202 Allocated 8192 byte pool (head: 7f602785a000, curr: 7f602785b020, alignment: 16)
[1687437627.239965 NOT 134] relay.c:2256 STARTUP - Relay has started...
[1687437627.240419 DBG 134] relay.c:2911 RINIT - Requests (1 total, 1 active, 1 max)
[1687437627.240557 DBG 134] relay.c:2878 HTTP_HOST - Failed to find HTTP_HOST or it's not a string.
[1687437627.240561 DBG 134] relay.c:2854 PLINK - Uninstalling handlers for 0 links
[1687437627.244063 DBG 136] shmalloc.c:202 Allocated 8192 byte pool (head: 7fe3905e5000, curr: 7fe3905e6020, alignment: 16)
[1687437627.244084 NOT 136] relay.c:2256 STARTUP - Relay has started...
[1687437627.244546 DBG 136] relay.c:2911 RINIT - Requests (1 total, 1 active, 1 max)
[1687437627.244632 DBG 136] relay.c:2878 HTTP_HOST - Failed to find HTTP_HOST or it's not a string.
[1687437627.244635 DBG 136] relay.c:2854 PLINK - Uninstalling handlers for 0 links
[1687437627.688526 DBG 128] cache.c:303 Failed to get global epoch
[1687437627.688536 DBG 128] relay.c:3003 RSHUTDOWN - Finished request
[1687437627.691781 NOT 128] relay.c:2279 Relay is shutting down, bye bye
[1687437627.767609 DBG 144] shmalloc.c:202 Allocated 8192 byte pool (head: 7f0b26001000, curr: 7f0b26002020, alignment: 16)
[1687437627.767635 NOT 144] relay.c:2256 STARTUP - Relay has started...
[1687437627.768184 DBG 144] relay.c:2911 RINIT - Requests (1 total, 1 active, 1 max)
[1687437627.768294 DBG 144] relay.c:2878 HTTP_HOST - Failed to find HTTP_HOST or it's not a string.
[1687437627.768297 DBG 144] relay.c:2854 PLINK - Uninstalling handlers for 0 links
[1687437628.218506 DBG 133] cache.c:303 Failed to get global epoch
[1687437628.218516 DBG 133] relay.c:3003 RSHUTDOWN - Finished request
[1687437628.223519 NOT 133] relay.c:2279 Relay is shutting down, bye bye
[1687437628.224956 DBG 131] cache.c:303 Failed to get global epoch
[1687437628.224964 DBG 131] relay.c:3003 RSHUTDOWN - Finished request
[1687437628.230701 NOT 131] relay.c:2279 Relay is shutting down, bye bye
[1687437628.230794 DBG 132] cache.c:303 Failed to get global epoch
[1687437628.230805 DBG 132] relay.c:3003 RSHUTDOWN - Finished request
[1687437628.237511 NOT 132] relay.c:2279 Relay is shutting down, bye bye
[1687437628.267620 DBG 148] shmalloc.c:202 Allocated 8192 byte pool (head: 7f6fbb4fe000, curr: 7f6fbb4ff020, alignment: 16)
[1687437628.267656 NOT 148] relay.c:2256 STARTUP - Relay has started...
[1687437628.268383 DBG 148] relay.c:2911 RINIT - Requests (1 total, 1 active, 1 max)
[1687437628.268516 DBG 148] relay.c:2878 HTTP_HOST - Failed to find HTTP_HOST or it's not a string.
[1687437628.268526 DBG 148] relay.c:2854 PLINK - Uninstalling handlers for 0 links
[1687437628.285346 DBG 149] shmalloc.c:202 Allocated 8192 byte pool (head: 7fa440640000, curr: 7fa440641020, alignment: 16)
[1687437628.285379 NOT 149] relay.c:2256 STARTUP - Relay has started...
[1687437628.286037 DBG 149] relay.c:2911 RINIT - Requests (1 total, 1 active, 1 max)
[1687437628.286157 DBG 149] relay.c:2878 HTTP_HOST - Failed to find HTTP_HOST or it's not a string.
[1687437628.286161 DBG 149] relay.c:2854 PLINK - Uninstalling handlers for 0 links
[1687437628.350519 DBG 122] cache.c:303 Failed to get global epoch
[1687437628.350538 DBG 122] relay.c:3003 RSHUTDOWN - Finished request
[1687437628.365093 NOT 122] relay.c:2279 Relay is shutting down, bye bye
```

## Environment

* Operating System and version: Debian
* PHP version: Happens on both 7.4 and 8.1

**Comment by dorrogeray** [comment]
**Update:** Setting `relay.maxmemory=0` is a workaround for the issue.

**Comment by michael-grunder** [comment]
Hi thanks for the report.

Is it possible to get a call stack from this?  It's going to be hard to track down without more context.  You could get one either by running your tests under a debugger which I understand may not be very feasible.  The other option is to generate a core dump and provide that along with your PHP binary.

**Comment by dorrogeray** [comment]
I have some updates. It turns out that the `relay.maxmemory=0` does not actually fix the problem, but `--passthru-php="-dzend.enable_gc=Off"` does. But this is a problem, since you need to set `zend.enable_gc=Off` to get a core dump. Seems like the only remaining path forward is to use xdebug profiler. I expect the profile file to be absolutely massive given the fact that it usually crashes somewhere in the middle of the integration tests, not sure if I will be able to produce it. I can run the scenario for you after few releases though, to see if the issue has gone away or not.

**Comment by dorrogeray** [comment]
Hi, any update on this - did the core dump provide the data needed to fix the issue?

**Comment by michael-grunder** [comment]
Hi @dorrogeray 

Unfortunately, the core dump seemed incomplete.  Do you know whether or not `ulimit -c` was `unlimited` when generating it?  I was able to inspect it a bit, but it wasn't in a state where I could glean much information.

If it is just that the core dump was truncated, generating another core file with `ulimit -c unlimited` might be helpful.  Alternatively what I could do is build a backtrace into Relay that would trigger on a SIGSEGV.

**Comment by tillkruss** [comment]
If you don't mind giving us access to your repo for a day, that would help solve this very quickly.

**Comment by dorrogeray** [comment]
@michael-grunder Yes I had the ulimit configured in the docker-compose configuration, but to be sure, I did re-run it again a double checked in the container with `ulimit -c` returning `unlimited`. I have also installed some debug symbols, resulting in debug symbol output like this:
![image](https://github.com/cachewerk/relay/assets/4999006/0f6aae53-bbe5-48a0-8e23-d1f112e9011b)
I have noticed that in gdb there are 2 threads using the `info threads` command, and I have switched to the second thread using `thread 2` command.
![image](https://github.com/cachewerk/relay/assets/4999006/f30faa2d-55e2-4b73-bfb0-318b149d07cf)

It seems that its related to crypto:
![image](https://github.com/cachewerk/relay/assets/4999006/02d0eb89-d9b5-45ce-b117-6b4329fc4fc2)

I will share this new core dump with you via email.

As far as docker images are concerned, we have chained build process so extracting it to a single file I could share is non trivial to say the least.

> Alternatively what I could do is build a backtrace into Relay that would trigger on a SIGSEGV.
I like this approach the most, if I understand this correctly, this would log / print some backtrace on SIGSEGV, regardless of where it happened in the php process?

@tillkruss I cannot give access to the repository, but I could schedule some call with screenshare.

**Comment by michael-grunder** [comment]
Awesome, thanks for the update.  I will take a look this weekend.

Can you do me a favor and post the output of the following command:

```bash
ldd /full/path/to/your/php/binary

# This also works if the php binary is the default one on your path
ldd $(which php)
```

The fact that it's related to libssl might mean this is simply a linker error.

**Comment by dorrogeray** [comment]
```sh
root@2be4e5d0e64b:/var/www/html# ldd /usr/bin/php7.4              
	linux-vdso.so.1 (0x00007ffe4c79b000)
	libargon2.so.1 => /usr/lib/x86_64-linux-gnu/libargon2.so.1 (0x00007f7510932000)
	libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007f7510918000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f75107d4000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f75107ce000)
	libxml2.so.2 => /usr/lib/x86_64-linux-gnu/libxml2.so.2 (0x00007f751061e000)
	libssl.so.1.1 => /usr/lib/x86_64-linux-gnu/libssl.so.1.1 (0x00007f751058b000)
	libcrypto.so.1.1 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007f7510295000)
	libpcre2-8.so.0 => /usr/lib/x86_64-linux-gnu/libpcre2-8.so.0 (0x00007f75101f7000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f75101da000)
	libsodium.so.23 => /usr/lib/x86_64-linux-gnu/libsodium.so.23 (0x00007f7510180000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f750ffac000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f750ff8a000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f7510df2000)
	libicuuc.so.67 => /usr/lib/x86_64-linux-gnu/libicuuc.so.67 (0x00007f750fd9f000)
	liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f750fd77000)
	libicudata.so.67 => /usr/lib/x86_64-linux-gnu/libicudata.so.67 (0x00007f750e25e000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f750e091000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f750e077000)
```

**Comment by michael-grunder** [comment]
What about `ldd /path/to/relay.so`?

**Comment by dorrogeray** [comment]
```sh
root@2be4e5d0e64b:/var/www/html# ldd /usr/lib/php/20190902/relay.so
	linux-vdso.so.1 (0x00007ffe21cf9000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f2103905000)
	libssl.so.1.1 => /usr/lib/x86_64-linux-gnu/libssl.so.1.1 (0x00007f2103872000)
	libcrypto.so.1.1 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007f210357e000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f21033aa000)
	libzstd.so.1 => /usr/lib/x86_64-linux-gnu/libzstd.so.1 (0x00007f21032cf000)
	liblz4.so.1 => /usr/lib/x86_64-linux-gnu/liblz4.so.1 (0x00007f21032ac000)
	libanl.so.1 => /lib/x86_64-linux-gnu/libanl.so.1 (0x00007f21032a4000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f2103bdd000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f210329e000)
```

**Comment by tillkruss** [comment]
Closing for now. If this isn't resolved in v0.7.0 feel free to re-open.
